pkg/controller/storagecluster/reconcile.go

```diff
@@ -172,7 +179,12 @@ func (r *ReconcileStorageCluster) Reconcile(request reconcile.Request) (reconcil
 		if phaseErr != nil {
 			reqLogger.Error(phaseErr, "Failed to set PhaseDeleting")
 		}
+
 		if contains(instance.GetFinalizers(), storageClusterFinalizer) {
+			err = r.setRookCleanupPolicy(instance, reqLogger)
```

Comment:
This propagates the cleanup policy in the delete part of the reconcile loop.
As discussed earlier today, at least as long as we don't have a finalizer for the CephCluster, this is kind of racy, and we should probably consider (additionall) propagating the policy as part of the normal reconcile code as well, to make it safer. I'd argue that we want the propagation always anyway, not only in delete as it's part of the storagecluster cr.

Reply:
@leseb The rook documentation says that the CleanupPolicy is to be set only when the deletion is imminent. https://github.com/rook/rook/blob/master/pkg/apis/ceph.rook.io/v1/types.go#L108

@obnoxxx and I were thinking of updating the rook CleanupPolicy in the main reconcile loop, even when the deletion is not imminent.

What do you suggest?

Reply:
I think it's fine to leave it only in the delete path for now. In particular, we have to consider the scenario where the StorageCluster is labeled and then deleted before ocs-operator reconciles it. This means that you would anyway have a CephCluster without the appropriate cleanup policy that you would need to amend. The cleanup policy itself means nothing prior to the CephCluster being marked for deletion, so updating it just prior to that deletion seems fine. I agree that it's a little racy, but that raciness won't be meaningfully mitigated by calling this logic in the standard reconciliation path.

Reply:
Just an FYI and it's also mention in Rook's doc, once the policy is in place all the future reconciles will be skipped.